src/transformers/models/emu3/modular_emu3.py

```diff
@@ -0,0 +1,1982 @@
+        sequence_length: int,
+        target_length: int,
+        dtype: torch.dtype,
+        device: torch.device,
+        cache_position: torch.Tensor,
+        batch_size: int,
+        **kwargs,
+    ):
+        """
+        Creates a causal 4D mask of shape `(batch_size, 1, query_length, key_value_length)` from a 2D mask of shape
+        `(batch_size, key_value_length)`, or if the input `attention_mask` is already 4D, do nothing.
+
+        Args:
+            attention_mask (`torch.Tensor`):
+                A 2D attention mask of shape `(batch_size, key_value_length)` or a 4D attention mask of shape
+                `(batch_size, 1, query_length, key_value_length)`.
+            sequence_length (`int`):
+                The sequence length being processed.
+            target_length (`int`):
+                The target length: when generating with static cache, the mask should be as long as the static cache,
+                to account for the 0 padding, the part of the cache that is not filled yet.
+            dtype (`torch.dtype`):
+                The dtype to use for the 4D attention mask.
+            device (`torch.device`):
+                The device to plcae the 4D attention mask on.
+            cache_position (`torch.Tensor`):
+                Indices depicting the position of the input sequence tokens in the sequence.
+            batch_size (`torch.Tensor`):
+                Batch size.
+        """
+        if attention_mask is not None and attention_mask.dim() == 4:
+            # In this case we assume that the mask comes already in inverted form and requires no inversion or slicing.
+            causal_mask = attention_mask
+        else:
+            min_dtype = torch.finfo(dtype).min
+            causal_mask = torch.full(
+                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
+            )
+            if sequence_length != 1:
+                causal_mask = torch.triu(causal_mask, diagonal=1)
+            causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+            causal_mask = causal_mask[None, None, :, :].expand(batch_size, 1, -1, -1)
+            if attention_mask is not None:
+                causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+                mask_length = attention_mask.shape[-1]
+                padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :]
+                padding_mask = padding_mask == 0
+                causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
+                    padding_mask, min_dtype
+                )
+
+        return causal_mask
+
+
+@add_start_docstrings(
+    "Emu3 Model with a head on top used for outputting logits for next token prediction.",
+    EMU3_START_DOCSTRING,
+)
+class Emu3ForCausalLM(Emu3PreTrainedModel, GenerationMixin):
+    config_class = Emu3TextConfig
```

Comment:
Same comment here, the new modular version should allow you to inherit from `LLamaForCausalLM` here (or better `GemmaForCausalLM` to avoid having the `pretraining_tp` part 🤗 maybe we can wait a few days until it's merged to see?

Reply:
oke, we can wait, In current situation inheriting from Llama added some extra modules, like Emu3TextAttention etc. Can I check the Emu3 modular on your branch if it is ready?

Reply:
Yes you can try with the current branch state! Lmk the result!

Reply:
Cool, I just tried it out. Awesome work, it fixes some stuff I was struggling with.

Some feedback on the current state. The modular for some reason copied `TextConfig` and `VQConfig` into modeling file also, instead of importing it `from .configuration_emu import ....`. 

Another thing I wanted to ask is if I can take say llama class and only change the inputs docstring name? For ex we have "LlamaInputDoc" but for new model we need "ModelTextInputDoc". I can overwrite the whole forward but I 'm not sure if I can specify only docstring for that

Reply:
Thanks for the feedback! Then I really need to cross-check imports when adding the classes, no workaround possible 😅 You want to simply change the decorator I suppose then? If so you can do it simply with:

```
@new_decorator(...)
def forward(**super_kwargs):
    super().forward()
```

`**super_kwargs` is a special keyword for modular which means "unpack all the args from the parent here". Then, `super().forward()` means "unpack the function definition from the parent here" (`super()` will also check for `del` statements and `self.xxx` statements immediately after it, and modify the code using those instead of the parent's, but here we don't need it)